sdk/servicebus/Microsoft.Azure.ServiceBus/src/Primitives/ConcurrentExpiringSet.cs

```diff
@@ -4,64 +4,90 @@
     using System.Collections.Concurrent;
+    using System.Threading;
     using System.Threading.Tasks;
 
     sealed class ConcurrentExpiringSet<TKey>
     {
         readonly ConcurrentDictionary<TKey, DateTime> dictionary;
-        readonly object cleanupSynObject = new object();
-        bool cleanupScheduled;
-        static TimeSpan delayBetweenCleanups = TimeSpan.FromSeconds(30);
+        readonly ICollection<KeyValuePair<TKey, DateTime>> dictionaryAsCollection;
+        readonly CancellationTokenSource tokenSource = new CancellationTokenSource();
+        volatile TaskCompletionSource<bool> cleanupTaskCompletionSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+        volatile int closeSignaled;
+        bool closed;
+        static readonly TimeSpan delayBetweenCleanups = TimeSpan.FromSeconds(30);
 
         public ConcurrentExpiringSet()
         {
             this.dictionary = new ConcurrentDictionary<TKey, DateTime>();
+            this.dictionaryAsCollection = dictionary;
+            _ = CollectExpiredEntriesAsync(tokenSource.Token);
         }
 
         public void AddOrUpdate(TKey key, DateTime expiration)
         {
+            this.ThrowIfClosed();
+
             this.dictionary[key] = expiration;
-            this.ScheduleCleanup();
+            this.cleanupTaskCompletionSource.TrySetResult(true);
         }
 
         public bool Contains(TKey key)
         {
+            this.ThrowIfClosed();
+
             return this.dictionary.TryGetValue(key, out var expiration) && expiration > DateTime.UtcNow;
         }
 
-        void ScheduleCleanup()
+        public void Close()
         {
-            lock (this.cleanupSynObject)
+            if (Interlocked.Exchange(ref this.closeSignaled, 1) != 0)
             {
-                if (this.cleanupScheduled || this.dictionary.Count <= 0)
-                {
-                    return;
-                }
-
-                this.cleanupScheduled = true;
-                Task.Run(async () => await this.CollectExpiredEntriesAsync().ConfigureAwait(false));
+                return;
             }
+
+            this.closed = true;
+
+            this.tokenSource.Cancel();
+            this.cleanupTaskCompletionSource.TrySetCanceled();
+            this.dictionary.Clear();
+            this.tokenSource.Dispose();
         }
 
-        async Task CollectExpiredEntriesAsync()
+        async Task CollectExpiredEntriesAsync(CancellationToken token)
         {
-            await Task.Delay(delayBetweenCleanups);
-
-            lock (this.cleanupSynObject)
+            while (!token.IsCancellationRequested)
             {
-                this.cleanupScheduled = false;
-            }
+                try
+                {
+                    await cleanupTaskCompletionSource.Task.ConfigureAwait(false);
+                    this.cleanupTaskCompletionSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
```

Comment:
This will not work. 
Add a `[Ignore]` test case temporarily and add an entry into dictionary with an expiration of 1 minute (greater than `delayBetweenCleanups` time) and see if it actually gets cleaned up..

Reply:
Instead, what you could do is, create a new `TaskCompletionSource` only when the dictionary is Empty. Make sure  concurrent calls to `AddOrUpdate` doesn't leave anything behind.

Reply:
Is it worth acquiring the locks on the concurrent dictionary to do the count or isempty check? I would doubt that

Reply:
If you have better ways of fixing it, sure, go for it! :)

Reply:
IsEmpty has a lock-free optimization starting with .NET Core 3.0 https://github.com/dotnet/corefx/pull/30098

Reply:
Pushed a proposal

Reply:
It passes this test

```
        [Fact]
        public void Contains_returns_false_for_expired_entry2()
        {
            var set = new ConcurrentExpiringSet<string>();
            set.AddOrUpdate("testKey", DateTime.UtcNow + TimeSpan.FromSeconds(61));

            Thread.Sleep(65000);

            Assert.False(set.Contains("testKey"), "The set should return false for an expired entry.");
        }
```

Reply:
Alternatively, if you want I could make the delayBetweenCleanups configurable and then add a test that simulates a short cleanup and an expiry that requires multiple cleanup cycles

Reply:
Then I could write something like this (still ugly though):

```

        [Fact]
        public void Contains_returns_false_for_expired_entry_that_requires_multiple_cleanup_cycles()
        {
            var set = new ConcurrentExpiringSet<string>(delayBetweenCleanups: TimeSpan.FromMilliseconds(50));
            set.AddOrUpdate("testKey", DateTime.UtcNow + TimeSpan.FromMilliseconds(200));

            Thread.Sleep(300);

            Assert.False(set.Contains("testKey"), "The set should return false for an expired entry.");
        }
```

Reply:
The test `Assert.False(set.Contains("testKey"), "The set should return false for an expired entry.");` would not be the complete test since the `Contains` method checks for actual expiration date on the value before returning. We needed to test if the item is in fact removed or not.. 
Either way, I think it should work right now with the `isEmpty` flag.